crates/ra_hir_ty/src/expr.rs

```diff
@@ -138,6 +148,70 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         }
     }
 
+    fn validate_call(&mut self, db: &dyn HirDatabase, call_id: ExprId, expr: &Expr) -> Option<()> {
+        // Check that the number of arguments matches the number of parameters.
+
+        // Due to shortcomings in the current type system implementation, only emit this diagnostic
+        // if there are no type mismatches in the containing function.
+        if self.infer.type_mismatches.iter().next().is_some() {
+            return Some(());
+        }
+
+        let is_method_call;
+        let (callee, args) = match expr {
+            Expr::Call { callee, args } => {
+                is_method_call = false;
+                let callee = &self.infer.type_of_expr[*callee];
+                let (callable, _) = callee.as_callable()?;
+                let callee = match callable {
+                    CallableDef::FunctionId(func) => func,
+
+                    // FIXME: Handle tuple struct/variant constructor calls.
+                    _ => return None,
+                };
+
+                (callee, args.clone())
+            }
+            Expr::MethodCall { receiver, args, .. } => {
+                is_method_call = true;
+                let callee = self.infer.method_resolution(call_id)?;
+                let mut args = args.clone();
+                args.insert(0, *receiver);
+                (callee, args)
+            }
+            _ => return None,
+        };
+
+        let loc = callee.lookup(db.upcast());
+        let ast = loc.source(db.upcast());
+        let params = ast.value.param_list()?;
+
+        let mut param_count = params.params().count();
+        let mut arg_count = args.len();
+
+        if params.self_param().is_some() {
```

Comment:
Urgh, we need to already invent some idiom for two param counts, there are many instances of `+1` code